README.md
# Adjustable Spool Holder

Spool holder which adapts to different widths and diameters of
filament spools while providing a smooth rolling experience on all
cases.

## Bill of Materials

P/N     | Description | Quantity 
-------:|:----------- |:--------:
290101  | Screw       | 1		
291101  | Clamp       | 2		
291201  | Nut         | 1
6905-ZZ | Bearing     | 2

This holder is designed to use bearings of size 42x25x9
(6905-xx). Same bearings but with different shields/seals
(i.e. 6905-2RS) should work too. For this type of application any
bearing quality would do.

## Print Settings
### Screw (2910)
- Rafts: No
- Supports: Yes
- Resolution:
  - <= 0.1 mm on the coupling ramp (start Z: 2.8 mm, finish Z: 3.8 mm)
  - <= 0.1 mm on the threaded section (start Z: 41.0 mm, finish Z: 106.6 mm)
  - Any elsewhere (I used 0.2 mm)
- Infill: >= 50%
- Vertical Shells: 3

### Clamp (2911)
- Rafts: No
- Supports: No
- Resolution: Any (I used 0.2 mm)
- Infill: >= 50%
- Vertical Shells: 3

### Nut (2912)
- Rafts: No
- Supports: No
- Resolution: <= 0.1 mm
- Infill: >= 20%
- Vertical Shells: 3

## Customise / Compile / Build
### Build STL files

Generate the intermediate SCAD file with the following command. Replace
<...> with the relevant file/name on each case. With no other parameters
this will generate a file called `<source_file>.scad`.
```
python <source_file>.py
```

Generate the STL file with:
```
openscad -o <stl_file>.stl <source_file>.scad
```

### Requirements
- [SolidPython][6]: a python front-end for solid modelling to OpenSCAD
- [OpenSCAD][7]: a solid 3D CAD modeller
- [Custom MCAD][8]: customised version of the MCAD library ('more-dev' branch)
- The [Open Sans][9] font (especially the "Bold" style)

[6]: https://github.com/SolidCode/SolidPython
[7]: http://www.openscad.org
[8]: https://github.com/rockstorm101/MCAD/tree/more-dev 
[9]: https://www.1001freefonts.com/open-sans.font

## Assembly

See images on `doc/images`. Orientation of the Clamps and the Nut are
critical. Clamps contact the bearings on their outer race and the
Screw and Nut contact the bearings on their inner race. It is critical
that these contacts occur on opposite sides of the bearings.

## Credits

The end coupling to be able to use the original nut from an Ender 3
spool holder is taken from [Ender 3 spool holder][1] by [@emolan][2].

[1]: https://www.thingiverse.com/thing:3979065
[2]: https://www.thingiverse.com/emolan

The font used for engraving is [Open Sans][9] by Steve Matteson.


## License
![open source hardware][3]
[![CC BY-SA 4.0][5]][4]

```
Copyright (c) 2021 Rock Storm

This work is open source hardware: you can redistribute it and/or modify
it under the terms of the Attribution-ShareAlike 4.0 International.

The code generating the models is distributed under the GNU General
Public License as published by the Free Software Foundation, either
version 3 of the License, or (at your option) any later version.

This work is distributed in the hope that it will be useful, but WITHOUT
ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
FITNESS FOR A PARTICULAR PURPOSE.
```

[3]: https://licensebuttons.net/l/by-sa/4.0/88x31.png
[4]: https://creativecommons.org/licenses/by-sa/4.0/
[5]: https://i1.wp.com/www.oshwa.org/wp-content/uploads/2014/03/oshw-logo-100-px.png


NEWS.md
# Adjustable Spool Holder - 0.1-p6 | Thread Fit Test
 - Thread clearances: -0.5 on male and +0.5 on female.
 - Thread printed at 0.1 layer height.
 - Using version 2.5 of threads file

# Adjustable Spool Holder - 0.1-p5 | Thread Fit Test
# Adjustable Spool Holder - 0.1-p4 | Thread Fit Test
# 0.1-p3 | Thread Fit Test

# Adjustable Spool Holder - 0.1-p2 | Coupling Fit Test
 - Ramp printed a 0.1 layer height
 - Supports everywhere

# Adjustable Spool Holder - 0.1-p1 | Thread Fit Test
 - Using version 1.3 of threads file



src/a2900.py
# Python imports
import solid as sp
import solid.utils as spu

# Import parts to be assembled
import p2910 as screw
import p2911 as clamp
import p2912 as nut
import v2901 as bearing

# SCAD imports
assembly = sp.import_scad('../lib/MCAD/assembly/attach.scad')

def part(variant = 'A', configuration = 'default', debug = False):
    S = sp.color("Khaki", 1)(screw.part(debug))
    B = bearing.part(debug)
    C = sp.color("SeaGreen", 1)( clamp.part(debug) )
    N = sp.color("Plum", 1)( nut.part(debug) )

    tmp = S

    tmp2 = B
    tmp2 += assembly.attach(bearing.RACEO_CONN, clamp.BEARING_CONN)(C)

    tmp += assembly.attach(screw.BEARING_CONN, bearing.RACEI_CONN)(tmp2)

    tmp3 = N
    tmp3 += assembly.attach(nut.BEARING_CONN, bearing.RACEI_CONN)(tmp2)

    tmp += assembly.attach(screw.NUT_CONN, nut.THREAD_CONN)(tmp3)

    tmp = sp.rotate([0,90,0])(tmp)
    if debug:
        tmp -= sp.translate([50-0.1,-50,0])(
            sp.color("White",0.1)( sp.cube([100,100,100], center = True) )
        )
    return tmp

if __name__ == '__main__':
    sp.scad_render_to_file(part(), include_orig_code = False)
    # One line per variant to generate all .scad files
    # sp.scad_render_to_file(part(variant = B), include_orig_code = False)


src/p2911.py
# Python imports
import math
import solid as sp
import solid.utils as spu

import v2901 as bearing

# SCAD imports
assembly = sp.import_scad('../lib/MCAD/assembly/attach.scad')
threads = sp.import_scad('../lib/MCAD/fasteners/threads.scad')

# Config
dim10 = 80    # OD
dim11 = bearing.OD + 0.2  # bearing seat ID
dim12 = 11.8  # bearing seat depth
dim13 = 15    # full height
dim14 = 38.5  # flange ID

# Connectors
DEFAULT_CONN = [[0, 0, 0], [0, 0, -1], 0]
BEARING_CONN = [[0, 0, dim12], [0, 0, -1], 0]

@spu.bom_part(description = "Clamp")
def part(variant = 'A', configuration = 'default', debug = False):
    tmp = sp.cylinder(h=dim13, d1=dim10, d2=dim14)
    tmp -= sp.translate([0,0,-0.1])(sp.cylinder(d=dim11, h=dim12+0.1))
    tmp -= sp.translate([0,0,-0.1])(sp.cylinder(d=dim14, h=dim13+0.2))
    if debug: tmp += assembly.connector(DEFAULT_CONN)
    return tmp

if __name__ == '__main__':
    sp.scad_render_to_file(part(), include_orig_code = False)
    # One line per variant to generate all .scad files
    # sp.scad_render_to_file(part(variant = B), include_orig_code = False)


src/p2912.py
# Python imports
import math
import solid as sp
import solid.utils as spu

# Project imports
import p2910 as screw

# SCAD imports
assembly = sp.import_scad('../lib/MCAD/assembly/attach.scad')
threads = sp.import_scad('../lib/MCAD/fasteners/threads.scad')
metric_fasteners = sp.import_scad('../lib/MCAD/fasteners/metric_fastners.scad')

# Config
T_d = screw.T_d  # thread nominal diameter
T_p = screw.T_p  # thread pitch
T_c = screw.T_c  # thread clearance
F_d = screw.F_d  # flange diam
F_t = 2          # flange thickness
B_d = screw.B_d  # bearing seat diameter
B_h = screw.B_h  # bearing seat depth
dim14 = F_t + B_h  # full length

# Connectors
THREAD_CONN = [[0, 0, 0], [0, 0, 1], 0]
BEARING_CONN = [[0, 0, F_t], [0, 0, 1], 0]

@spu.bom_part(description = "Nut")
def part(variant = 'A', configuration = 'default', debug = False):
    tmp = sp.cylinder(d=F_d, h=F_t)
    tmp += sp.translate([0,0,F_t-0.1])(sp.cylinder(d=B_d, h=B_h+0.1))
    thread = threads.metric_thread(diameter = T_d,
                                   pitch = T_p,
                                   length = dim14+0.2,
                                   internal = True,
                                   clearance = T_c)
    chamfer = threads.chamfer_cylinder(T_d-2*T_p, T_d+0.5, internal = True)
    thread = threads.chamfered_thread(dim14+0.2, internal = True)(
        thread, chamfer, chamfer)
    tmp -= sp.translate([0,0,-0.1])(thread)
    # tmp -= chamfer_cylinder_internal(T_d+0.25)
    # tmp -= sp.translate([0,0,dim14])(
    #     sp.rotate([180,0,0])(
    #         chamfer_cylinder_internal(T_d+0.25) ) )
    if debug: tmp += assembly.connector(BEARING_CONN)
    return tmp

def chamfer_cylinder_internal(diameter = 10, angle = 45):
    d = diameter/2
    h = d / math.tan(math.radians(angle))
    h_p = h + 0.1
    d_p = h_p / h * d
    tmp = sp.translate([0,0,-0.1])(sp.cylinder(h = h_p, d1 = 2*d_p, d2 = 0))
    return tmp

if __name__ == '__main__':
    sp.scad_render_to_file(part(), include_orig_code = False)
    # One line per variant to generate all .scad files
    # sp.scad_render_to_file(part(variant = B), include_orig_code = False)


src/p2910.py
# Python imports
import math
import solid as sp
import solid.utils as spu

import v2901 as bearing

# SCAD imports
assembly = sp.import_scad('../lib/MCAD/assembly/attach.scad')
threads = sp.import_scad('../lib/MCAD/fasteners/threads.scad')

# Config
dim10 = 15    # first step height
B_h = bearing.B + 0.4   # bearing seat height
F_d = 29.5  # flange OD, which the bearing seats against
B_d = bearing.ID - 0.2  # bearing seat diam
dim14 = 90    # full length. TODO: bring back to 90
T_d = 20      # thread nom diam
T_p = 2.5     # thread pitch
T_c = 0.4     # thread clearance
dim17 = dim14 - dim10 - B_h  # thread length


# Connectors
BEARING_CONN = [[0, 0, dim10], [0, 0, 1], 0]
NUT_CONN =     [[0, 0, dim14-dim17/2+0.9], [0, 0, -1], 0]

@spu.bom_part(description = "Screw")
def part(variant = 'A', configuration = 'default', debug = False):
    tmp = sp.cylinder(d=F_d, h=dim10)
    tmp += sp.translate([0,0,dim10-0.1])(sp.cylinder(d=B_d, h=B_h+0.1))
    #tmp += sp.translate([0,0,+0.1])(sp.cylinder(d=T_p, h=dim14-0.1))
    thread = threads.chamfered_thread(dim17, internal = False)(
        threads.metric_thread(diameter = T_d,
                              pitch = T_p,
                              length = dim17,
                              internal = False,
                              clearance = 1 ),
        threads.chamfer_cylinder(T_d, T_d - 2*T_p - 0.5, internal = False)
    )
    tmp += sp.translate([0,0,dim17+B_h+dim10])( sp.rotate([180,0,0])(thread) )
    # tmp -= sp.translate([0,0,dim14])(
    #     sp.rotate([180,0,0])(
    #         chamfer_cylinder_external(2, 45, diameter = 20) ))
    tmp += sp.rotate([180,0,0])(
        sp.translate([-107,-113,-0.2])(
            sp.import_(
                "../aux/Ender_3_spool_holder/files/" +
                "Ender3_Spool_Holder_Coupling.stl") ) )
    if debug: tmp += assembly.connector(DEFAULT_CONN)
    return tmp

def chamfer_cylinder_external(length = 1, angle = 45, diameter = 10):
    # create a solid chamfer that then has to be substracted
    #   length = chamfer length
    #   angle  = chamfer angle
    #   diameter = diameter of the cylinder onto which this chamfer
    #              will be applied
    e = length
    tg_a = math.tan(math.radians(angle))
    d = diameter
    h = e / tg_a
    e_p = e + 0.1*(1 + tg_a)
    h_p = h + 0.1*(1 + 1/tg_a)
    p1 = [d/2 + 0.1 - e_p,      -0.1, 0]
    p2 = [d/2 + 0.1,            -0.1, 0]
    p3 = [d/2 + 0.1,       h_p - 0.1, 0]
    tmp = sp.rotate_extrude()( sp.polygon([p1, p2, p3]) )
    return tmp

if __name__ == '__main__':
    sp.scad_render_to_file(part(), include_orig_code = False)
    # One line per variant to generate all .scad files
    # sp.scad_render_to_file(part(variant = B), include_orig_code = False)
